v-next/hardhat/src/internal/builtin-plugins/solidity-test/task-action.ts

```diff
@@ -49,6 +49,11 @@ const runSolidityTests: NewTaskActionFunction<TestActionArguments> = async (
     return;
   }
 
+  // NOTE: We run the compile task first to ensure all the artifacts for them are generated
+  // Then, we compile just the test sources. We don't do it in one go because the user
+  // is likely to use different compilation options for the tests and the sources.
+  await hre.tasks.getTask("compile").run();
```

Comment:
I think we need to reintroduce the `--no-compile` here, otherwise running `hardhat test` will compile twice.